Help: introduce InlineHelp that uses a Popover to keep users in context

lsinger opened the pull request:
This PR introduces an Inline Help system to Calypso. Previously, clicking the `(?)` at the bottom of the sidebar would directly link to `/help`. We now open a Popover that displays the top four help articles (taken from `/help`) and has basic search functionality. There's also a link to `/help`. 

Current status:

![help popover](https://user-images.githubusercontent.com/191598/35351752-db791514-010f-11e8-8dc3-8096b9431063.gif)

To test:
- Make sure the Popover looks and works well. 
- Make sure the search looks and works well in different situation — when not searching, during search, when finding something, when finding nothing. 
- Make sure the link to `/help` works correctly. 
- Make sure help results with special characters in their title are rendered correctly (e.g. search for "top posts"). 
- Make sure this works the same in different sections of Calypso. 
- Make sure this works and looks fine on mobile. 
- Make sure help can be opened (g, h) and closed (Esc) using the keyboard, and make sure you can navigate (arrow up / down) and select (Enter) search results. 
- Make sure the following interactions send off Tracks events: open / close the Inline Help, search for something, open a search result, open "more help."

shaunandrews commented:
<img width="375" alt="screen shot 2018-01-22 at 4 37 20 pm" src="https://user-images.githubusercontent.com/191598/35245834-8c8b94a0-ff92-11e7-8c41-598372cc74b6.png">

This is looking good. Updating a bunch of placeholders and CSS, and added some styles for an `.is-active` state on the sidebar footer's help button.

Each link opens the help doc in a new tab — I wonder if we should keep them in the same tab/window? If not, then we probably need to add an external icon for each result (which I kind of hate).

Any thoughts on adding some simple keyboard navigation for results? Up/down arrows to navigate, and enter to load the result.

lsinger commented:
@shaunandrews:

- using the `is-active` class now
- went back opening links in the same tab
- added keyboard navigation, use sequence "g", "h" ("go" to "help") to open help, the rest as you described — screen cap below

![keyboard-help](https://user-images.githubusercontent.com/23619/35282873-00a5b068-0057-11e8-9a02-a7baec5b61c0.gif)

lsinger commented:
@shaunandrews the button looks different when you're on `/help` — is that on purpose?

![help-button](https://user-images.githubusercontent.com/23619/35327922-1fbea8de-00fb-11e8-85b8-382ffcb45f02.gif)

shaunandrews commented:
> is that on purpose?

Nope, should be fixed. 🚢

frontdevde commented on client/blocks/inline-help/style.scss:
```diff
@@ -0,0 +1,237 @@
+
+.inline-help__results-placeholder {
+	margin: 24px 16px;
+	min-height: 188px - 48px;;
```
Minor: double semicolon

frontdevde commented on client/blocks/inline-help/style.scss:
```diff
@@ -0,0 +1,237 @@
+	}
+}
+
+.inline-help__heading {
```
Minor: iirc the coding guidelines for Calypso recommend to not indent deep using sass functionality but to list the selectors individually e.g.`.search-card .search` if the specificity is needed. 

Also, there are a lot of different absolute px values used. Wondering if some of those would benefit from storing them in variables (or deriving them from other variables) to make future maintenance/iterations easier.

Both not necessary, just thinking out loud.

jblz commented:
Loving this. 

One thing... scrolling up / down does weird things to the vertical placement.  Perhaps the element can be fixed differently somehow?

![scrolling-affects-fixed-position](https://user-images.githubusercontent.com/1587282/35352881-632540fc-0113-11e8-9b83-558db0bcd1d8.gif)

shaunandrews commented:
> scrolling up / down does weird things to the vertical placement. Perhaps the element can be fixed differently somehow?

Should be fixed now, @jblz — thanks!

mattsherman commented:
I gave this a quick spin... overall, I'm really excited to see this take shape. Nice work so far! 🎉

A few comments/issues:

- Opening a link in the same tab is frustrating. I read how to do something, and then need to go back in the browser, wait for Calypso to reload, and then try to remember what I just read. It would be nice if I could at least right click on a link and open in new tab (but, since these are actually buttons I can't).
- When using the keyboard, you never see the tooltips for the results. It would be nice to see that information without having to hover over a result. But, I can see how that could be text overload!
- If you search for something that has no results (I mistakenly typed "posrt"), the searching animation never ends
- When searching, the popover is slightly shorter than when with four results. This slight shift in height feels a bit unpolished.
- Maybe it's just me, but the focus outline on the search field extending past the popover looks odd
- When tabbing, the focus outlines on the results looks odd. I would expect the items to be highlighted like they do when you use the arrow keys to go through them.

<img width="349" alt="screencapture at wed jan 24 19 13 10 est 2018" src="https://user-images.githubusercontent.com/2098816/35364232-f8187990-013b-11e8-9872-c9d0a0d382da.png">

mtias commented on client/blocks/inline-help/style.scss:
```diff
@@ -0,0 +1,257 @@
+.sidebar .sidebar__footer .button,
```
Just a quick thing, but styles that target inline-help as a child of sidebar should be added to the sidebar component, layout component, etc. Only direct `inline-help` styles should be on this file.

mtias commented on client/blocks/inline-help/style.scss:
```diff
@@ -0,0 +1,257 @@
+			border-radius: 4px 4px 0 0;
+			height: auto;
+
+			&.is-open.has-focus {
```
Isn't all this already coming from the search component?

shaunandrews replied:
You would think, but its not.

mtias commented on client/blocks/inline-help/style.scss:
```diff
@@ -0,0 +1,257 @@
+			}
+
+			.search__input {
+				font-size: 18px;
```
Why do you have to touch font-sizes? There's a `compact` version of search if needed.

mtias commented on client/blocks/inline-help/style.scss:
```diff
@@ -0,0 +1,257 @@
+	}
+}
+
+@keyframes inline-help__results-placeholder-loading {
```
This looks like a good candidate to be made into an easy to use component `PlaceholderLines`

mtias commented on client/components/popover/index.jsx:
```diff
@@ -129,13 +135,18 @@ class Popover extends Component {
-		setTimeout( () => this.setPosition(), 0 );
+		if ( ! this.isUpdatingPosition ) {
+			this.debug( 'requesting to update position after render completes' );
+			requestAnimationFrame( () => {
```
What's the reasoning for this?

lsinger replied:
The reasoning is in the message for 38e779cfebf6dd42bb30d8be7127951b0c531ae2: 

```
Popover: update position even when own props stay the same
Previously, the Popover didn't update its position when its children changed. This would lead to a mis-positioned Popover when the height of its children was changed.

"This change makes sure that the Popover recalculates its position even when its props are the same. This makes sure that it is correctly positioned when only its children change. This change uses `this.isUpdatingPosition` to avoid infinite updating loops.```

mtias replied:
Might be worth including an inline comment on the code.

lsinger replied:
👍 Good call!

rachelmcr commented:
Since the inline help duplicates the search functionality that's on `/help`, should it still pop up when you're on `/help`? (Currently when you're at https://wordpress.com/help the help icon is highlighted and doesn't do anything.)

If we keep the inline help active on `/help`, we'll need to sort out the interaction that's happening between the search in the help popup and the main search box. Right now, changing the input in one search box also changes the results in the other section:

![inlinehelpsearch](https://user-images.githubusercontent.com/8658164/35382054-10c4e1ca-01b6-11e8-9919-4cc1e8731e2e.gif)

rachelmcr commented:
If I search, close the inline help, and then reopen the inline help, my search results are retained but the search field is cleared. Could we either retain the text in the search field or clear the search results?

Some of the search results end up centered. It looks like it's always the same support docs, no matter where they appear in the list of results:

![screenshot 2018-01-25 10 09 36](https://user-images.githubusercontent.com/8658164/35382731-0b8f9f22-01b8-11e8-9545-7de1c62532e9.png)

>Opening a link in the same tab is frustrating. I read how to do something, and then need to go back in the browser, wait for Calypso to reload, and then try to remember what I just read.

To add on to this, if I go back in the browser and want to look at a different search result (e.g. if the first result I selected didn't have the information I needed) I have to start the search over again because the previous search is cleared.

lsinger commented:
Thanks for your comments, everyone! ❤️ Addressed most review comments in 0fe55be. 

I can't reproduce the centered result, @rachelmcr. Do you have some more details regarding OS and browser? Pretty odd that it's just a single one. 😕 

One remaining issue I want to address for this PR is the confusing interplay of things when visiting `/help` and using Inline Help there.

rachelmcr commented:
>I can't reproduce the centered result, @rachelmcr. Do you have some more details regarding OS and browser? Pretty odd that it's just a single one.

I was seeing this in Mac OS 10.12.6 (Sierra), Chrome 63.0.3239.132.

However, when I went to go try it again and gather more details about what I was seeing, I couldn't load the calypso.live link anymore. I'm getting a blank Calypso screen with a console error:

![screenshot 2018-01-25 16 22 10](https://user-images.githubusercontent.com/8658164/35399352-f2ee9926-01eb-11e8-98cd-b975d7ab12ea.png)

lsinger commented:
@rachelmcr I have the same problem running it locally — might be related to a rebase I did to get in sync with `master` again. Looking into it.

lsinger commented:
@rachelmcr regarding the [weird interaction on `/help`](https://github.com/Automattic/wp-calypso/pull/21495#issuecomment-360417957) — I think I inadvertently fixed it while fixing something else. I can't reproduce anymore. is it fixed for you as well?

rachelmcr commented:
I'm no longer seeing the centered results for any of the links, so something must have fixed that. :)

However, I'm still seeing the weird interaction on `/help` with these steps:

1. At `/help`, search in the main search bar.
2. Open the inline help and search for a different keyword there. Note the results in the main search area change to match, but the search there doesn't change.
3. Remove the search from the inline help search box, and note that the main search area is left with placeholder search results.

lsinger commented:
@rachelmcr confirm I can reproduce. Thanks for the steps!

shaunandrews commented:
> Since the inline help duplicates the search functionality that's on /help, should it still pop up when you're on /help?

Yes, I think it should, at least for now. If `/help` continues to have a sidebar, then the pattern of "help in the bottom of the sidebar" should exist everywhere.

lsinger commented:
Whew! With 9a28b60 I switched this over from the `/help`-screen's flux store to its own shiny new redux store. 

@rachelmcr can you check again if you can find any problems still? Thanks so much! 

/cc @shaunandrews @mtias @jblz @mattsherman @frontdevde

marekhrabe commented:
- [ ] The sidebar footer help, even on /help makes sense. However, I think it shouldn't have an "active" state. Previously, it was active on that route as an indication the help it opened but now I think it should only be active once the popover is opened. 

<img width="300" alt="screen shot 2018-01-30 at 12 09 40" src="https://user-images.githubusercontent.com/156676/35563370-c079d126-05b6-11e8-9773-c86e28031df5.png">

- [ ] One tiny nitpick - when on "/help" and clicking "More help" in the popover, I think the popover should close.

Just a note about the "g h" keyboard shortcut, it is only active in development. Keyboard shortcuts were never activated in production (I don't know why)

lsinger commented:
> The sidebar footer help, even on /help makes sense. However, I think it shouldn't have an "active" state. Previously, it was active on that route as an indication the help it opened but now I think it should only be active once the popover is opened. 

For now I'd like to keep this as we have no other indication of what page we're on right now when on `/help`. That is, there's no other menu item or button that appears to be "active" — except the avatar in the Masterbar, which doesn't really indicate help. 

I think to solve this we'll need to rethink a few more things, but that doesn't seem in scope for this PR. 

> One tiny nitpick - when on "/help" and clicking "More help" in the popover, I think the popover should close.

Great catch! Fixed in 7843f2f.

> Just a note about the "g h" keyboard shortcut, it is only active in development. Keyboard shortcuts were never activated in production (I don't know why)

Huh — that's sad. I'll look into why (separately).

marekhrabe commented on client/blocks/inline-help/index.jsx:
```diff
@@ -0,0 +1,135 @@
+		searchResults: getInlineHelpSearchResultsForQuery( state, ownProps.searchQuery ),
+	};
+};
+const mapDispatchToProps = null;
```
I there is no need, I would remove this

lsinger replied:
👍 Removing.

marekhrabe commented on client/blocks/inline-help/inline-help-search-results.jsx:
```diff
@@ -0,0 +1,186 @@
+	componentWillUpdate( nextProps ) {
+		if ( nextProps.shouldOpenSelectedResult ) {
+			const url = this.getSelectedUrl();
+			this.followHelpLink( url );
```
Not sure what is this line trying to do. `this.followHelpLink` returns a function but it's not assigned to a variable here or called.

lsinger replied:
👍 Actually calling it.

marekhrabe commented on client/blocks/inline-help/inline-help-search-results.jsx:
```diff
@@ -0,0 +1,186 @@
+		const classes = [
+			'inline-help__results-item',
+			this.props.selectedResult === index && 'is-selected',
+		].filter( Boolean );
```
as this array goes to `classnames`, I would change it into an object like:
```
const classes = {
  'inline-help__results-item': true,
  'is-selected': this.props.selectedResult === index,
};
```

or have the first class assigned statically and the object would be just for the selection part:

```
const classes = { 'is-selected': this.props.selectedResult === index };

<li … className={ classNames( 'inline-help__results-item', classes ) }>
```

marekhrabe commented on client/blocks/inline-help/inline-help-search-results.jsx:
```diff
@@ -0,0 +1,186 @@
+	}
+}
+
+const mapStateToProps = ( state, ownProps ) => {
```
nice-to-have simplification for your `mapStateToProps` in this PR might be:

```
const mapStateToProps = ( state, ownProps ) => ( {
  searchResults: …,
  isSearching: …,
} );
```

saves a bit of brackets and additional nesting but I guess comes down to a personal preference. up to you to ignore me :)

marekhrabe commented on client/components/data/query-inline-help-search/index.jsx:
```diff
@@ -0,0 +1,49 @@
+}
+
+export default connect(
+	( state, ownProps ) => {
```
`( state, { query } ) => {` might save you the `const` part and would allow this to be simplified to just pure object return like I've mentioned in other comment above

lsinger replied:
Nice, good call! 👍

lsinger commented:
Thanks for the review, @marekhrabe! Addressed your comments in e7eeef8.

michaeldcain commented on client/blocks/inline-help/index.jsx:
```diff
@@ -0,0 +1,134 @@
+ * Internal Dependencies
+ */
+import config from 'config';
+import { tracks } from 'lib/analytics';
```
It's recommended to use the [analytics middleware](https://github.com/Automattic/wp-calypso/tree/master/client/state/analytics) instead of calling the analytics library directly. There's a basic example of the implementation [here](https://github.com/Automattic/wp-calypso/blob/4b1cde3f6a412505c0a1aee180abd61d6e9206d4/client/my-sites/people/people-invites/invites-list-end.jsx#L14).

michaeldcain commented on client/blocks/inline-help/index.jsx:
```diff
@@ -0,0 +1,134 @@
+
+	showInlineHelp = () => {
+		debug( 'showing inline help.' );
+		tracks.recordEvent( 'calypso_inline-help_show' );
```
I don't think that Tracks [event names](https://github.com/Automattic/wp-calypso/blob/master/client/lib/analytics/README.md#tracks-naming-conventions) support dashes.

michaeldcain commented on client/blocks/inline-help/index.jsx:
```diff
@@ -0,0 +1,134 @@
+						/>
+
+						<Button onClick={ this.moreHelpClicked } className="inline-help__button" borderless href="/help">
+							<Gridicon icon="help" /> More help
```
Missing `translate()`.

lsinger commented:
Thanks for the review, @michaeldcain! Addressed your comments in 9842621, 4a4e06b, and 0420ea5.

stephanethomas commented on assets/stylesheets/_main.scss:
```diff
@@ -54,6 +54,12 @@ body,
+	// Specifically, this makes clicking outside a popover work on touch devices. -shaun
+	@include breakpoint( "<660px" ) {
+		cursor: pointer;
+	}
```
@shaunandrews, @lsinger: this displays a wrong cursor when hovering elements that are not clickable on small viewports (I tested this on desktop only). Is that expected?

<img width="205" alt="screenshot" src="https://user-images.githubusercontent.com/594356/37829567-c5a7277c-2e9f-11e8-9536-e6ebca12fd9d.png">

lsinger replied:
Judging from the code: expected, maybe, but definitely not intended. The intention is to make clicking outside of the popover work. 

However, as viewports as small as that will usually not display a cursor anyway (touch screen devices, etc.), I'm tempted to leave this as is. Except if there's a better solution for the problem, of course, or we find that users with pointing devices but super narrow browser windows are a group we should support better.
